CLI Key Management

Philip-NLnetLabs opened the pull request:
Demonstrate key management using a CLI tool. Initial interface between key management and signer.

Philip-NLnetLabs committed:
Initial version of keyset.

tertsdiepraam commented:
This seems to be missing the file for the `keyset` command.

Philip-NLnetLabs committed:
Now with keyset.rs.

Philip-NLnetLabs committed:
Fix Cargo.toml

Philip-NLnetLabs committed:
The CDS/CDNSKEY RRsets also need to be signed. Cron command.

Philip-NLnetLabs committed:
Display *-generate-params as *-algorithm in show.

Philip-NLnetLabs committed:
Avoid starting a KSK keyroll before init.

Philip-NLnetLabs committed:
Make sure key tags are unique.

Philip-NLnetLabs committed:
Support for algorithm rolls.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Fix up merge.

Philip-NLnetLabs committed:
Remove unneeded format!.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Store key references are file URLs.

Philip-NLnetLabs committed:
More detailed error messages.

Philip-NLnetLabs committed:
More URL handling.

Philip-NLnetLabs committed:
Add support for cron_next.

Philip-NLnetLabs committed:
Use clap to parse keyset subcommands (#100)

Co-authored-by: Terts Diepraam <terts.diepraam@gmail.com>

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

mozzieongit reviewed (commented):
I know this is still a draft, but here my notes from reading this PR to understand the other PRs:

- Add comments on the subcommands, that say what they do and how to use them
  - And a note about what to do after each command? e.g. after init / creating key, mention which command should be called next and when
- Split up big run function into sub functions (obsolete by #108)
- Consistently use `env.in_cwd()` vs raw paths (e.g. `read_to_string(priv_url.path())`) or make clear why not using `env.in_cwd()`
- Add tests? I'm sure not all can be tested easily, but maybe some can
- Explain to users which TTL value is expected in the propagation/cache subcommands
- adjust the actions output after each command, to something human-actionable, so the user knows what to do with that information

ximon18 committed:
Add support for the KMIP cryptographic backend. (#99)

- Adds KMIP server based key generation, signing and destruction, equivalent to the existing Ring/OpenSSL functionality.
- Adds new kmip subcommands for managing KMIP server configurations.
- Adds support for referring to KMIP keys by a new KMIP URL scheme. 
- Add a feature for the KMIP crypto backend just like the Ring and OpenSSL crypto backends.
- Adds support for storing sensitive credentials in files separate to the KMIP server configuration.

Philip-NLnetLabs committed:
Add support for automatic key rolls (#108)

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Restore white space.

ximon18 committed:
Getting ready to publish an 0.1.0-rc2 version of dnst with keyset support to the proposed channel at packages.nlnetlabs.nl.

ximon18 committed:
Add missed Cargo.lock change due to rc2 version bump.

ximon18 committed:
Mark keyset as experimental.

ximon18 committed:
Extend O/S's supported for packaging.

ximon18 committed:
Test missing/incorrect O/S variants.

ximon18 committed:
Permit KMIP servers to be added in an inactive state. (#122)

ximon18 committed:
Delete errant character that broke the YML syntax.

Philip-NLnetLabs committed:
Keyset import (#121)

* Restructure roll commands.

* Import public keys.

* Import a public/private key pair from files.

* Add a default TTL to config. Use that for DNSKEY/CDS/CDNSKEY/DS RRsets.

* Cargo.lock.

* Support for importing KMIP keys.

* Import public/private keys in decoupled state

* Add --private-key option to importing a public/private key pair from files.

* Add remove-key command.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Update Cargo.lock.

Philip-NLnetLabs committed:
Keyset status (#124)

* Verbose status output.

* Add TODO list.

* Get rid of .map_err::<Error, _>. Just .map_err is fine if the .into() is removed.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Fallout from merging main: more map_err.

Philip-NLnetLabs committed:
Add support for imported keys to Init. (#126)

tertsdiepraam commented on src/commands/mod.rs:
```diff
@@ -94,10 +99,12 @@ impl Command {
             Self::Notify(notify) => notify.execute(env),
             Self::SignZone(signzone) => signzone.execute(env),
             Self::Update(update) => update.execute(env),
+            // Self::Help(help) => help.execute(env),
```
Nit but I think this stems from a conflict.

Philip-NLnetLabs replied:
As far as I know, help was disabled by Jannik until the help command can render man pages.

Philip-NLnetLabs replied:
Ah, I see, in commit 6915b816a65dbafa1ed27f0cd84781362dbb7f0d:
-            // Self::Help(help) => help.execute(env),
+            Self::LdnsUpdate(ldnsupdate) => ldnsupdate.execute(env),

That was just too confusing.

Philip-NLnetLabs committed:
Implement the set default-ttl command. (#127)

Philip-NLnetLabs committed:
Create parent directories for config and state files. (#129)

bal-e committed:
Update 'Cargo.lock'

bal-e committed:
[keyset] Don't fail on repeated 'init'

Cascade uses this because it can't easily tell whether the keyset state
is already initialized, in some cases.

ximon18 committed:
Replace `Keys(s)` with `key(s)`. (#130)

Philip-NLnetLabs committed:
Revert "[keyset] Don't fail on repeated 'init'"

This reverts commit 67b99a10a4d9ae61924192b272e09cbed9af93a7.

Philip-NLnetLabs committed:
Removed unused import.

Philip-NLnetLabs committed:
Keyset man (#128)

Co-authored-by: Ximon Eighteen <3304436+ximon18@users.noreply.github.com>

ximon18 committed:
Package as cascade-dnst instead of dnst. (#131)

To avoid Cascade installing dnst and uninstalling ldns-utils that Cascade users may want to use such as ldns-verify-zone, which dnst doesn't provide yet. Also because this branch of dnst depends on an as yet unreleased version of the domain crate.

Update Cargo.lock bcause cargo-deb complains that it needs updating.

Co-authored-by: Jannik Peters <jannik@nlnetlabs.nl>

ximon18 committed:
dnst keyset man page tweaks. (#133)

Co-authored-by: Jannik Peters <jannik@nlnetlabs.nl>
Co-authored-by: Philip-NLnetLabs <philip@nlnetlabs.nl>

mozzieongit committed:
Fix typo in docs

mozzieongit committed:
Fix error message saying .pub expected

ximon18 committed:
Typo correction.

ximon18 committed:
Minor typo correction in man page sources.

ximon18 committed:
Minor typo correction in man page sources.

ximon18 committed:
Fix RST syntax error in man page source.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Introduce a 'WorkSpace' object to keep the current working state of k… (#138)

* Introduce a 'WorkSpace' object to keep the current working state of keyset.

* Fallout from merging.

* Switch to the crypto-and-keyset-fixes branch in domain for the time being.
Disable kmip because the branch does not support that.

* Update lock file.

* Bump Rust version to 1.85 because domain is at 1.85.0.

* Switch to domain-kmip. (#142)

* Switch to domain-kmip.

* Use https URL for domain-kmip.

* Update for domain-kmip change.

* Switch to main branch of domain-kmip.

* Clippy.

* Bump Rust version to 1.88 for kmip-protocol.

Philip-NLnetLabs committed:
Undo Cascade-related changes.

Philip-NLnetLabs committed:
Update manual pages.

Philip-NLnetLabs committed:
Merge branch 'main' into keyset

Philip-NLnetLabs committed:
Update man pages.
